### PR TITLE
Add a --allow-branch-switch to fwupdmgr

### PR DESCRIPTION
**Pull request by hughsie**
This allows us to turn off the tests like version format differences and
checking for downgrades.

Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [X] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by superm1** [comment]
So we have both a `switch-branch` command and a `--allow-switch-branch` option?  It feels like it should be an either/or.

**Comment by hughsie** [comment]
> and a --allow-switch-branch option

From a user point of view, the only time they'd need to use `--allow-switch-branch option` is when installing a local .cab file. It also lets us add the functionality into fwuptool for testing in another commit.

**Comment by superm1** [comment]
> From a user point of view, the only time they'd need to use --allow-switch-branch option is when installing a local .cab file. It also lets us add the functionality into fwuptool for testing in another commit.

But so then maybe `switch-branch` should not be a command is my point.  It should be `install-release` and then --allow-switch-branch would show all the different releases including multiple branches.

**Comment by hughsie** [comment]
> But so then maybe switch-branch should not be a command is my point

Not so convinced; from a 40,000ft view the CLI and API between downgrading and switching a release should be the same; they're both actions that have security implications and don't fit in the "JFDI update" flow. You could argue we don't need downgrade as a CLI command, as you could just have `install-release` and get the user to select the one that has the same branch as existing and that doesn’t have is_upgrade or doesn't have is_downgrade. The point is the user explicitly wants to "switch the branch" to a different stream rather than specifically install a release on that branch.

That said, maybe `install-release [DEVICE_ID] [VERSION] [BRANCH] [REMOTE]` is a useful thing to have, in addition. Just not on this PR :)

**Comment by superm1** [comment]
But so then how is --allow-switch-branch different than force?

**Comment by hughsie** [comment]
> how is --allow-switch-branch different than force

`--force is so much more powerful` -- insomuch that it will ignore the verfmt (like --allow-branch-switch) but **also** checks like "is the binary blob the right size" and checking CRCs and checksums are correct. I'm actually wondering if we want to allow `--force` from the daemon at all, and leave it as something only supported in fwpdtool.

**Comment by superm1** [comment]
> I'm actually wondering if we want to allow --force from the daemon at all, and leave it as something only supported in fwpdtool.

We've always left `--force` as the big hammer to overrride any specific checks.  It's useful even in `fwupdmgr`, for example if you want to disable upower plugin checks.

Maybe what would make sense is to instead split up force into a variety of individual types of force?

**Comment by hughsie** [comment]
> split up force into a variety of individual types of force

Yes, I think so too. Will ponder.

**Comment by hughsie** [comment]
@superm1 what do you think about ^? that means that you can only ignore failed CRC and VID/PID mismatch from fwupdtool, but I think that's actually what we want. `--force` then becomes a _gentle nudge_ rather than a _potentially-hardware-destroying_ action.

File: src/fu-tool.c [hunk 1]
```diff
@@ -2422,6 +2424,12 @@ main (int argc, char *argv[])
 		{ "force", '\0', 0, G_OPTION_ARG_NONE, &force,
 			/* TRANSLATORS: command line option */
 			_("Override plugin warning"), NULL },
+		{ "ignore-checksum", '\0', 0, G_OPTION_ARG_NONE, &ignore_checksum,
+			/* TRANSLATORS: command line option */
+			_("Ignore firmware checksum failures"), NULL },
+		{ "ignore-vid-pid", '\0', 0, G_OPTION_ARG_NONE, &ignore_vid_pid,
```
Comment (superm1) [inline src/fu-tool.c:2430]: these should be added to bash completion

**Comment by superm1** [comment]
> @superm1 what do you think about ^? that means that you can only ignore failed CRC and VID/PID mismatch from fwupdtool, but I think that's actually what we want. `--force` then becomes a _gentle nudge_ rather than a _potentially-hardware-destroying_ action.

I think yeah that's the right direction.  We probably should completely remove `--force` though, and everything it did should be it's own thing, and some things only doable from `fwupdtool`.  IE also have `--ignore-power`

File: src/fu-tool.c (hunk 1 above)
Comment (hughsie) [reply src/fu-tool.c:2430]: I think there should be some kind of swear-jar for every time I forget to add these. My apologies.

**Comment by hughsie** [comment]
>  We probably should completely remove --force though

Isn't that part of our CLI, API and ABI? I'd be fine slightly redefining `--force` to be a "soft nudge" that's not going to cause damage. I'm not super-keen on removing the API or emulating it.

**Comment by superm1** [comment]
> I'd be fine slightly redefining --force to be a "soft nudge" that's not going to cause damage

Well I was thinking marking it as adding `--force` shows a message `--force` is removed, please see these other commands for the same things.

**Comment by superm1** [comment]
> to be a "soft nudge

Or maybe rename it `-ignore-version-format` but let `--force` be a synonym for it?

**Comment by hughsie** [comment]
I added `FWUPD_INSTALL_FLAG_IGNORE_POWER` but I'm not sure we want/need a `_IGNORE_FLASHES_LEFT` and `IGNORE_UEFI_DBX_ESP` and `_IGNORE_VERSION_FORMAT`. The latter is really just a _the uploader screwed something up_ and now we can rewrite the metainfo.xml file on the LVFS it's easiest just to fix the problem IMHO.

File: src/fu-util.c [hunk 2]
```diff
@@ -3003,10 +3012,16 @@ main (int argc, char *argv[])
 		priv->flags |= FWUPD_INSTALL_FLAG_ALLOW_REINSTALL;
 	if (allow_older)
 		priv->flags |= FWUPD_INSTALL_FLAG_ALLOW_OLDER;
-	if (force)
+	if (allow_branch_switch)
+		priv->flags |= FWUPD_INSTALL_FLAG_ALLOW_BRANCH_SWITCH;
+	if (force) {
 		priv->flags |= FWUPD_INSTALL_FLAG_FORCE;
+		priv->flags |= FWUPD_INSTALL_FLAG_IGNORE_POWER;
```
Comment (superm1) [inline src/fu-util.c:3019]: this same behavior adjustment seems missing on `fu-tool.c`

Comment (hughsie) [reply src/fu-util.c:3019]: Indeed, good catch.

Comment (hughsie) [reply src/fu-util.c:3019]: Done.

File: src/fu-util.c [hunk 3]
```diff
@@ -2011,6 +2011,7 @@ fu_util_switch_branch (FuUtilPrivate *priv, gchar **values, GError **error)
 	g_signal_connect (priv->client, "device-changed",
 			  G_CALLBACK (fu_util_update_device_changed_cb), priv);
 	priv->flags |= FWUPD_INSTALL_FLAG_ALLOW_REINSTALL;
+	priv->flags |= FWUPD_INSTALL_FLAG_ALLOW_BRANCH_SWITCH;
```
Comment (superm1) [inline src/fu-util.c:2014]: I guess it's worth asking - why not also have this same argument on `fwupdtool`?  Installing a CAB could encounter the same situation.

File: src/fu-util.c [hunk 4]
```diff
@@ -2683,6 +2686,9 @@ main (int argc, char *argv[])
 		{ "allow-older", '\0', 0, G_OPTION_ARG_NONE, &allow_older,
 			/* TRANSLATORS: command line option */
 			_("Allow downgrading firmware versions"), NULL },
+		{ "allow-branch-switch", '\0', 0, G_OPTION_ARG_NONE, &allow_branch_switch,
+			/* TRANSLATORS: command line option */
+			_("Allow switching firmware branch"), NULL },
 		{ "force", '\0', 0, G_OPTION_ARG_NONE, &force,
 			/* TRANSLATORS: command line option */
 			_("Override warnings and force the action"), NULL },
```
Comment (superm1) [inline src/fu-util.c:2694]: This should probably change to reflect it's been gimped.

Comment (hughsie) [reply src/fu-util.c:2694]: I changed this to `Force the action by relaxing some runtime checks`

File: src/fu-util.c (hunk 3 above)
Comment (hughsie) [reply src/fu-util.c:2014]: You're completely correct. Fixed.

**Review by superm1 — APPROVED** [review]

**Pull request merged**